src/main/java/com/group6/harmoniq/controllers/SpotifyController.java

```diff
@@ -0,0 +1,157 @@
+import org.springframework.web.util.UriComponentsBuilder;
+import org.springframework.ui.Model;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+
+import jakarta.servlet.http.Cookie;
+import jakarta.servlet.http.HttpServletResponse;
+
+
+@Controller
+public class SpotifyController {
+    
+    String client_id = System.getenv("SPOTIFY_CLIENT_ID");
+    String client_secret = System.getenv("SPOTIFY_CLIENT_SECRET");
+    String redirect_uri = "https://harmoniq-1.onrender.com/callback";
+    final String stateKey = "spotify_auth_state";
+    
+    
+    //redirects to Spotify API's Authorization endpoint
+   @GetMapping("/login")
+   public String login(HttpServletResponse response) {
+    SecureRandom secureRandom;
+    try {
+        secureRandom = SecureRandom.getInstanceStrong();
+    } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException(e);
+        }
+
+        String state = generateRandomString(secureRandom, 16);
+        Cookie stateCookie = new Cookie(stateKey, state);
+        response.addCookie(stateCookie);
+        String scope = "user-read-private user-read-email user-follow-read";
+
+        //Requesting Authorization
+        String authUrl = UriComponentsBuilder.fromHttpUrl("https://accounts.spotify.com/authorize")
+            .queryParam("response_type", "code")
+            .queryParam("client_id", client_id)
+            .queryParam("scope", scope)
+            .queryParam("redirect_uri", redirect_uri)
+            .queryParam("state", state)
+            .toUriString();
+
+        return "redirect:" + authUrl;
+   }
+   
+    //Checks application state
+    //Requests an access token using user client_id and client_secret
+    @GetMapping("/callback")
+    public String callback(@RequestParam("code") String code, @RequestParam("state") String state, @CookieValue(stateKey) String storedState,
+    HttpServletResponse response, Model model) {
+
+        if (state == null || !state.equals(storedState)) {
+            return "redirect:/login.html";
+        } else {
+       
+            response.addCookie(new Cookie(stateKey, null));      
+            String tokenUrl = "https://accounts.spotify.com/api/token";
+            String authHeader = "Basic " + Base64.getEncoder().encodeToString((client_id + ":" + client_secret).getBytes(StandardCharsets.UTF_8));
+
+            Map<String, String> bodyParams = new HashMap<>();
+            bodyParams.put("code", code);
+            bodyParams.put("redirect_uri", redirect_uri);
+            bodyParams.put("grant_type", "authorization_code");
+        
+            RestTemplate restTemplate = new RestTemplate();
+            HttpHeaders headers = new HttpHeaders();
+            headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);
+            headers.set("Authorization", authHeader);
+       
+            MultiValueMap<String, String> map = new LinkedMultiValueMap<>();
```

Comment:
Maybe rename it to something more meaningful like requestBody so its more clear
```suggestion
            MultiValueMap<String, String> requestBody = new LinkedMultiValueMap<>();
```